#! /bin/sh /usr/share/dpatch/dpatch-run
## 100503_r27337_rb_string_value_cstr.dpatch by Daigo Moriwaki <daigo@debian.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: back-ported from the upstream r27337
## DP: string.c (rb_string_value_cstr): make NUL terminated if it is not done.

@DPATCH@
Index: b/string.c
===================================================================
--- a/string.c	2010-07-20 21:45:44.000000000 +0900
+++ b/string.c	2010-07-20 21:46:18.000000000 +0900
@@ -1265,10 +1265,12 @@
 {
     VALUE str = rb_string_value(ptr);
     char *s = RSTRING_PTR(str);
+    long len = RSTRING_LEN(str);
 
-    if (!s || RSTRING_LEN(str) != strlen(s)) {
+    if (!s || memchr(s, 0, len)) {
 	rb_raise(rb_eArgError, "string contains null byte");
     }
+    if (s[len]) rb_str_modify(str);
     return s;
 }
 
